File: AdventOfCode/2018/AdventOfCode/Day01.cs
using System.Collections.Generic;

namespace AdventOfCode
{
    public static class Day01
    {
        public static int Part1(string input)
        {
            var result = 0;

            foreach (var line in input.Split("\n".ToCharArray()))
            {
                var value = int.Parse(line.Substring(1));

                if (line[0] == '+')
                {
                    result += value;
                }
                else
                {
                    result -= value;
                }
            }

            return result;
        }

        public static int Part2(string input)
        {
            var result = 0;
            var seen = new HashSet<int>();
            var index = 0;
            var lines = input.Split("\n".ToCharArray());

            while (true)
            {
                var value = int.Parse(lines[index].Substring(1));

                if (lines[index][0] == '+')
                {
                    result += value;
                }
                else
                {
                    result -= value;
                }

                if (seen.Contains(result))
                {
                    return result;
                }

                seen.Add(result);

                index++;
                index %= lines.Length;
            }
        }
    }
}

File: AdventOfCode/2024/day01.py
from collections import Counter

def part1():
    lines = [line.strip() for line in open("day01.txt").readlines()]
    col1 = []
    col2 = []

    for line in lines:
        one, two = line.split()
        col1.append(int(one))
        col2.append(int(two))

    col1 = sorted(col1)
    col2 = sorted(col2)

    result = 0

    for i in range(len(col1)):
        result += abs(col2[i] - col1[i])

    print(result)

def part2():
    lines = [line.strip() for line in open("day01.txt").readlines()]
    col1 = []
    col2 = []

    for line in lines:
        one, two = line.split()
        col1.append(int(one))
        col2.append(int(two))

    col1 = sorted(col1)
    col2 = Counter(col2)

    similarity_score = 0

    for id in col1:
        similarity_score += id * col2[id]

    print(similarity_score)


def main():
    part1()
    part2()


if __name__ == "__main__":
    main()
